File: java-start/app/src/main/java/com/vonage/tutorial/messaging/ChatViewModel.java
package com.vonage.tutorial.messaging;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.nexmo.client.*;
import com.nexmo.client.request_listener.NexmoApiError;
import com.nexmo.client.request_listener.NexmoRequestListener;

import java.util.ArrayList;

public class ChatViewModel extends ViewModel {

    private NexmoClient client = NexmoClient.get();

    private NexmoConversation conversation;

    private MutableLiveData<String> _errorMessage = new MutableLiveData<>();
    LiveData<String> errorMessage = _errorMessage;

    private MutableLiveData<String> _userName = new MutableLiveData<>();
    LiveData<String> userName = _userName;

    private MutableLiveData<ArrayList<NexmoEvent>> _conversationEvents = new MutableLiveData<>();
    ;
    LiveData<ArrayList<NexmoEvent>> conversationEvents = _conversationEvents;

    private NexmoMessageEventListener messageListener = new NexmoMessageEventListener() {
        @Override
        public void onTextEvent(@NonNull NexmoTextEvent textEvent) {
            updateConversation(textEvent);
        }

        @Override
        public void onAttachmentEvent(@NonNull NexmoAttachmentEvent attachmentEvent) {

        }

        @Override
        public void onEventDeleted(@NonNull NexmoDeletedEvent deletedEvent) {

        }

        @Override
        public void onSeenReceipt(@NonNull NexmoSeenEvent seenEvent) {

        }

        @Override
        public void onDeliveredReceipt(@NonNull NexmoDeliveredEvent deliveredEvent) {

        }

        @Override
        public void onTypingEvent(@NonNull NexmoTypingEvent typingEvent) {

        }
    };

    public void onInit() {
        getConversation();
        _userName.postValue(client.getUser().getName());
    }

    private void getConversation() {
        // TODO Fetch the conversation
    }


    private void getConversationEvents(NexmoConversation conversation) {
        // TODO: Fetch the conversation events
    }

    private void updateConversation(NexmoTextEvent textEvent) {
        ArrayList<NexmoEvent> events = _conversationEvents.getValue();

        if (events == null) {
            events = new ArrayList<>();
        }

        events.add(textEvent);
        _conversationEvents.postValue(events);
    }

    public void onSendMessage(String message) {
        // TODO: end new message to client SDK
    }

    public void onBackPressed() {
        client.logout();
    }

    public void onLogout() {
        client.logout();
    }

    @Override
    protected void onCleared() {
        // TODO: Unregister message listener
    }
}

File: kotlin-start/app/src/main/java/com/vonage/tutorial/messaging/ChatViewModel.kt
package com.vonage.tutorial.messaging

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.nexmo.client.NexmoAttachmentEvent
import com.nexmo.client.NexmoClient
import com.nexmo.client.NexmoConversation
import com.nexmo.client.NexmoDeletedEvent
import com.nexmo.client.NexmoDeliveredEvent
import com.nexmo.client.NexmoEvent
import com.nexmo.client.NexmoMessageEventListener
import com.nexmo.client.NexmoSeenEvent
import com.nexmo.client.NexmoTextEvent
import com.nexmo.client.NexmoTypingEvent
import com.vonage.tutorial.messaging.extension.toLiveData

class ChatViewModel : ViewModel() {

    private val client: NexmoClient = NexmoClient.get()

    private var conversation: NexmoConversation? = null

    private val _errorMessage = MutableLiveData<String>()
    val errorMessage = _errorMessage.toLiveData()

    private val _userName = MutableLiveData<String>()
    val userName = _userName.toLiveData()

    private val _conversationEvents = MutableLiveData<List<NexmoEvent>?>()
    val conversationEvents = _conversationEvents.toLiveData()

    private val messageListener = object : NexmoMessageEventListener {
        override fun onTypingEvent(typingEvent: NexmoTypingEvent) {}

        override fun onAttachmentEvent(attachmentEvent: NexmoAttachmentEvent) {}

        override fun onTextEvent(textEvent: NexmoTextEvent) {
            // TODO: Update the conversation
        }

        override fun onSeenReceipt(seenEvent: NexmoSeenEvent) {}

        override fun onEventDeleted(deletedEvent: NexmoDeletedEvent) {}

        override fun onDeliveredReceipt(deliveredEvent: NexmoDeliveredEvent) {}
    }

    fun onInit() {
        getConversation()
        _userName.postValue(client.user.name)
    }

    private fun getConversation() {
        TODO("Fetch the conversation")
    }

    private fun getConversationEvents(conversation: NexmoConversation) {
        TODO("Fetch the conversation events")
    }

    private fun updateConversation(textEvent: NexmoEvent) {
        val events = _conversationEvents.value?.toMutableList() ?: mutableListOf()
        events.add(textEvent)
        _conversationEvents.postValue(events)
    }

    fun onSendMessage(message: String) {
        TODO("Send new message to client SDK")
    }

    fun onBackPressed() {
        client.logout()
    }

    fun onLogout() {
        client.logout()
    }

    override fun onCleared() {
        TODO("Unregister message listener")
    }
}
